## Appstore Data Analysis for Forecasting Prospective Apps for Increased Ads' Revenue

* This project is supposed to get a better insight on which apps to be developed both for android and apple market.The basic intention is to make free apps and get revenue through ads,so by analysing the data we may tell what kind of apps to be developed for greater revenue by increasing users

* The business idea runs as to find profitable app categories from the android market,develop a beta of app,upload it to
  android store,If the app does good in terms of installs and reviews ,then build a advanced version and monitor its     revenue for six months.If the revenues are healthy,then develop the similar app for ios appstore

* The datasets for both android play store and ios store are obtained from the open source data science platform Keggle.
  Links for both data sets are mentioned below
  
1. Android play store [dataset](https://www.kaggle.com/lava18/google-play-store-apps/home)
2. IOS appstore [dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home)
  
* This project exolores various methods for manipulating list,dictionaries and frequency tables.Apart from that,the major emphasis of this project is explore the dataset with limited manual commands and automize the whole process.

* Throughout the project google playstore ,android store and other well known terms used for android market points to **Google Playstore**.Similarly **apple appstore** and **ios appstore** are used equivalently

## Function for opening Dataset

* named as **open_dataset(data_set,header= False)**
* opens Dataset as list of a list
* separates header which is basically the information of different attributes
* takes in by default header parameter to be true.If there is no header in dataset,to be mentioned explicilty

In [1]:
def open_dataset(data_set,header=True):
    opened = open(data_set)
    from csv import reader
    read_file = reader(opened)
    data = list(read_file)
    data_set = data[1:]
    attr_info = data[0]
    return data_set ,attr_info



In [2]:
apple,apple_head = open_dataset('/home/bashir/Documents/Dataquest/Guided Project: Profitable App Profiles for the App Store and Google Play Markets/app-store-apple-data-set-10k-apps/AppleStore.csv')

In [3]:
google,google_head = open_dataset('/home/bashir/Documents/Dataquest/Guided Project: Profitable App Profiles for the App Store and Google Play Markets/google-play-store-apps/googleplaystore.csv')

## explore_data() function

* takes in dataset without header,starting index,end index
* print out the information as specified
* as an optional feature,if row_and _columns are made * True * ,also print out number of features and total rows(apps_data) in the data set

In [4]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

## Attributes Information

* Next step naturally is to get some information about the attributes i.e.column names in the data set
* For google play store,they are stored in google_head
* For apple play store,they are stored in apple_head

#                  Impotrant Attributes from Playstore

* The following attributes will give a good insight in making best selling apps for android market place

| **Attributes**     | **Information**                |
|:-----------    |:---------------------------|
| **App** | Name of the app
| **Rating**         | Average user rating        |
| **Price**          | price category             |
| **Installs**       | Total downloads            |
| **Genres**         | Category the app falls into|
| **Content Rating** | Suitability for age groups |

* Further information about the data set can be obtained by clicking the following [link](https://www.kaggle.com/lava18/google-play-store-apps/home)

* Below we can view ,attributes names and some rows of the sata set

In [5]:
print(google_head)
print('----------------------------------------------------------------')
print('Some rows or apps contained :')
print('----------------------------------------------------------------')
explore_data(google,0,3)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
----------------------------------------------------------------
Some rows or apps contained :
----------------------------------------------------------------
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']




#                  Impotrant Attributes from Appstore

* The following attributes will give a good insight in making best selling apps for ios market place

| Attributes             | Information                |
|:-------------------    |:---------------------------|
| **user_rating**        | Average user rating        |
| **price**              | Price category             |
| **track_name**         | App name                   |
| **prime_Genre**        | Category the app falls into|
| **cont_Rating**       | Suitability for age groups |
| **size_bytes**         | App size in bytes          |

* Further information about the data set can be obtained by clicking the following [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home)

*  Below we can view ,attributes names and some rows of the sata set

In [6]:
print(apple_head)
print('----------------------------------------------------------------')
print('Some rows or apps contained :')
print('----------------------------------------------------------------')
explore_data(apple,0,3)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
----------------------------------------------------------------
Some rows or apps contained :
----------------------------------------------------------------
['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']




## Anamolities in Google Store Data

* One row has a missin value for categories and in return other values are also in wrong order.This error is discussed in great detail here [error](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015).

* The best possible solution is to locate the row and delete it because one row wont affect the analysis

* Below we find the index of the row from our dataset and then will delete it

In [7]:
''' finding the exact location from over dataset which does 
    not contain the header.It is 10473 row so its index is 
    10472'''

explore_data(google,10472,10473) 

# deleting the specified row
del google[10472]

# using the same index again to see if its deleted
explore_data(google,10472,10473)


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']




## Duplicate Entries in Google Play store Data

* First approach of any data  science technique is to look out for duplicate data.
* The google playstore data set contains duplicate data which can be studied in the following keggle discussion
   [duplicate_entries](https://www.kaggle.com/lava18/google-play-store-apps/discussion/67894)
* But its always better to look for yourself,for which we created a frequency table i.e. dictionery which stores the names of apps and count their repetitions.
* Since the data is crumbersome, we made it even more interesting by storing the key of the dictioneries i.e. apps name having duplicate entries.In this way,by just printing out the few entries of list,we can observe repetitions.
* Total duplications are also kept in record which can be observed from the repetitions variable
* For better understanding a frequency table is also maintained to check repetition against every app
* check_for_duplicate performs this task and it is completely generic because it requires a dataset in form of rows of rows and index containing the names.    

In [8]:
def check_for_duplicate(dataset,name_index):
    duplicate_freq ={}
    list_duplicate = []
    repetition = 0
    for row in dataset:
        name = row[name_index]
        if name in duplicate_freq:
            duplicate_freq[name] += 1
            repetition +=1
        else:
            duplicate_freq[name] = 1


    for key in duplicate_freq:
        if duplicate_freq[key] > 1:
            list_duplicate.append(key)


    return  repetition,list_duplicate, duplicate_freq




    


## Results and Observation

* By observing the data set using the routine we have 1181 repetitive entries
* Some of the apps are shown which are duplicated.
* Lets study the data for 'Mcqueen Coloring pages' and make some decisions on how to handle duplication



In [9]:
repetition,list_duplicate,frequency = check_for_duplicate(google,0)

print('Total repetitive entries in google data set are :', repetition)
print('------------------------------------------------------------------')
print('Some of the  repetitive entries are')
print(list_duplicate[:5])
print('---------------------------------------------')
print('Number of duplicate entries for "Mcqueen Coloring pages":')
print(frequency['Mcqueen Coloring pages' ])

Total repetitive entries in google data set are : 1181
------------------------------------------------------------------
Some of the  repetitive entries are
['Coloring book moana', 'Mcqueen Coloring pages', 'UNICORN - Color By Number & Pixel Art Coloring', 'Textgram - write on photos', 'Wattpad 📖 Free Books']
---------------------------------------------
Number of duplicate entries for "Mcqueen Coloring pages":
2


## Comparing the Duplicate Entries

* Lets compare the duplicate entries and see how can we come up with a system to get rid of them and make the data cleaner

In [10]:
 for row in google:
        if row[0]== 'Mcqueen Coloring pages':
            print(row)
            print('------------------------------')

['Mcqueen Coloring pages', 'ART_AND_DESIGN', 'NaN', '61', '7.0M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Action & Adventure', 'March 7, 2018', '1.0.0', '4.1 and up']
------------------------------
['Mcqueen Coloring pages', 'FAMILY', 'NaN', '65', '7.0M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Action & Adventure', 'March 7, 2018', '1.0.0', '4.1 and up']
------------------------------


#### By close Inspection we can observe,the differences lie in:

* Categories
* Number of reviews
* We can infer that since we have genre as an indicator in the dataset also which matches for both entries and a better indicator than category which is a bigger set.And difference in number of reviews means that with time,some users also decided to give their opinion about the app whether googd or bad.
* The practical decision would be to keep the row i.e appdata with greater number of reviews

## Steps in Cleaning the Duplicate entries from Google Dataset

1. As discussed above,we need to take the app copy which has the maximum number of reviews.For this reason,we create a 
   dictionery **reviews_max** and store that entries of apps having maximum reviews.This dictionery will help us in       creating our non duplicate data set **android_clean**
  
2. To create the dctionery,loop through the *google* dataset,store the app name as name and review as n_reviews.Check if the name is in the dictionery as a key and whether the value at that key is less than n_reviews.If both conditions are true ,store the n_reviews as updated value.In this way,we update the review for each app in the dictionery.

3. Now its time to create the dateset without duplicate entries i.e. android_clean.For that puporse create two empty lists,**android_clean** and **already_added**. 

4. Loop through the **google dataset**,store the app name as name and review as n_reviews again.Check for the name to be not part of the **already_added list**  and  **n_reviews** matches the value at the corresponding key of  **reviews_max**,then append the whole row to the **android_clean** list(In this way,we get the list of list dataset just like **google**) and store the name in **already_added** list

5. If the name is already in the **already_added** ,we will do nothing,in this way our new data set will be free of duplicate entries.

6. Now its time to check,if the cleaned data is actually valid.Its length should be equal to length of google minus **repetition**


In [11]:
# Creating dictionery containing only single entries of apps sorted with their maximum reviews

reviews_max = {}
for row in google:
    name = row[0]
    n_reviews = float(row[3])
    if (name in reviews_max) and reviews_max[name]< n_reviews :
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
       
    

In [12]:
android_clean =[]  # new android dataset without any duplicates
already_added =[]  #book_keeping list for apps names that are already in non<-duplicate version of dataset
for row in google:
    name = row[0]
    n_reviews = float(row[3])
    if (n_reviews == reviews_max[name]) and name not in already_added:
        android_clean.append(row)
        already_added.append(name)

In [13]:
# This block will print true if the cleaned data has correct length
check_length = len(google) - repetition
print('If the new cleaned dataset is valid ?\n',check_length == len(android_clean))

If the new cleaned dataset is valid ?
 True


## Analyzing Apps with English Names

* At this point of our analysis,we are interested in only analyzing apps for the english speaking market.
* For this reason,we have to remove all the **Non-English** apps from the data.
* We will employ the use of ASCII characters.Characters in english language have ASCII Values upto 127.
* But for some apps including **'Docs To Go™ Free Office Suite'** and **'Instachat😜'**,characters like **™** and **😜** have ASCII values greater than 127 but we dont want to remove our legitimate data.

* For this reason, we keep a check that if the app name has under three characters greater than 127, we consider it to be a English name

* Function **is_thename_english** checks for the condition

In [14]:
def is_thename_english(string):
    count = 0
    for c in string:
        
        if ord(c) > 127:
            count +=1
            if count > 3:
                return False
     
    return True

## Google Playstore Dataset with English Names

1. Below ,we remove the apps having other than english names.
2. **english_android_clean** is the new polished dataset.
3. Basically we move through each row of our dataset,apply the function **is_thename_english**, and append the correct row to *englih_android_clean*,making it a list of  a list.
4. We can observe the number of non-english apps also by   ***len(android_clean)-len(english_android_clean)***

In [15]:
# english_android_clean is new datatset contained for english apps
english_android_clean =[]
for row in android_clean:
   
    name = row[0]
    if (is_thename_english(name)):
        english_android_clean.append(row)
print('The total number of English language apps are :') 
print(len(english_android_clean))
print('--------------------------------------------------')
print('The total number of non-English language apps are :')
print(len(android_clean)-len(english_android_clean))

The total number of English language apps are :
9614
--------------------------------------------------
The total number of non-English language apps are :
45


## Apple Playstore Dataset with English Names

1. Now applying the same procedure with the apple dataset.
2. Just to keep in mind that the index for app names is 1 instead of 0
3. **english_apple** is the new dataset containing only english name apps
4. There are 6183 apps suitable for English speaking market

In [16]:
# english_apple is new datatset contained for english apps
english_apple =[]
for row in apple:
   
    name = row[2]
    if (is_thename_english(name)):
        english_apple.append(row)
print('The total number of English language apps are :') 
print(len(english_apple))
print('--------------------------------------------------')
print('The total number of non English language apps are :')
print(len(apple)-len(english_apple))

The total number of English language apps are :
6183
--------------------------------------------------
The total number of non English language apps are :
1014


## Free Apps

* The final step of our data cleaning is to sort the datasets with respect to only free apps.
* Since it was our main goal to predict type of free apps that would be installed more and in return give us revenue through ads.So for this reason our final dataset for android will be **android_final** and for apple ios will be **apple_final**


In [17]:

# first column of apple al
apple_final = []
for row in english_apple:
    price = float(row[5])
    if price == 0:
        apple_final.append(row)

# for google play data set
android_final = []
for row in english_android_clean:
    price = float(row[7].replace('$',''))
    if price == 0:
        android_final.append(row)
        
print('After all cleaning')
print('------------------------------------------------------------------------------')
print('The Android Appstore(Google Appstore) contains entries :', len(android_final))
print('And attributes(columns)',len(google_head))
print('------------------------------------------------------------------------------')
print('The IOS Appstore(Apple Appstore) contains entries :', len(apple_final))
print('And attributes(columns)',len(apple_head))
        


After all cleaning
------------------------------------------------------------------------------
The Android Appstore(Google Appstore) contains entries : 8864
And attributes(columns) 13
------------------------------------------------------------------------------
The IOS Appstore(Apple Appstore) contains entries : 3222
And attributes(columns) 17


# Business Model for App Development

We can explain the business model we are interested in as follows: 

1. Make an android app (minimal and lite version) and upload it on google store.
2. If it gets a good response from users in terms of downloads of the beta,develop it full and optimize it.
3. Wait for 6 months,if the app turns profitable ,devlop an ios version also.

* We can analyze our datasets and get insights for popular apps by creating a frequnce table wrt genre.Lets begin



## What is  'freq_generator'  function doing

1. It takes in a dataset i.e. list of a list and the index of the dataset for which it builds a frequency table.
2. Showing the frequency table in percentage is easier to visualize for which the **len** function comes in handy.
3. We are also interested in sorting the frequency table in descending order. The function **sorted()** does not work that efficiently with dictionaries i.e. it only considers its keys,not their corresponding values.
4. For that specific reason, we convert the each key-value pair to a tuple.It is important to place the value first.
5. Append these tupples to a list and now the **sorted()** function will give us the required insight.


In [18]:
def freq_generator(dataset,index):
    genre_freq = {}
    freq_list = []
    for row in dataset:
        genre_var = row[index]
        if genre_var in genre_freq:
            genre_freq[genre_var] += 1
        else :
            genre_freq[genre_var] = 1

    for genre_var in genre_freq:
        genre_freq[genre_var] /= len(dataset)
        genre_freq[genre_var] *= 100

   

    for key in genre_freq:
        t = (genre_freq[key],key)
        freq_list.append(t)
    
    
    freq_list = sorted(freq_list,reverse=True)
        
    return freq_list
    
   
    

    

# Google Data Set


### Contribution from top 20 Categories and Genre App

* After the frequency table is set up, we are interested to see the frequencies for *Category* and *Genres* columns.
* It would also help to print out the contributions from top 20 categories and genres in the list.
* This analysis would help us to focus on probable most downloaded categories and genres and save the hassle to look into unimportant one

### ______________The  Case Of Categories___________________________

In [19]:

print('The following are valuable insights from the "Category" column of Google play store dataset :')
google_Cat_freq = freq_generator(android_final,1)

# Check for top 20 Categories contributions

slice_freq = google_Cat_freq[:20]
top_20 = 0
print('----------------------------------------------------------------')
print('Below are the top_20 Categories of most Downloaded apps with their percentage :')
for item in slice_freq:
    top_20 += float(item[0])
    print(item[1],':',item[0])

print('-----------------------------------------------------------------')
print('The top_20 apps Categories contribute to Downloads :')
print(top_20,'%')



The following are valuable insights from the "Category" column of Google play store dataset :
----------------------------------------------------------------
Below are the top_20 Categories of most Downloaded apps with their percentage :
FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
-----------------------------------------------------------------
The top_20 apps Categories contribute to Downloads :
88.52

### ______________The  Case Of Genres___________________________

In [20]:

print('The following are valuable insights from the "Genre" column of Google play store dataset :')
google_Genres_freq = freq_generator(android_final,9)

# Check for top 20 Categories contributions

slice_freq = google_Genres_freq[:20]
top_20 = 0
print('----------------------------------------------------------------')
print('Below are the top_20 Genres of most Downloaded apps with their percentage  :')
for item in slice_freq:
    top_20 += float(item[0])
    print(item[1],':',item[0])

print('-----------------------------------------------------------------')
print('The top_20 apps Genres contribute to Downloads :')
print(top_20,'%')


The following are valuable insights from the "Genre" column of Google play store dataset :
----------------------------------------------------------------
Below are the top_20 Genres of most Downloaded apps with their percentage  :
Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
-----------------------------------------------------------------
The top_20 apps Genres contribute to Downloads :
72.8339350180

## Takeaways from Android category and genre frequency table

1. The **Genre** column seems to be a  subset of **Category** column until that point of our analysis and is more rational to find patterns from **Genre Column** for in-depth analysis.

2. There is no stark difference in genres unlike ios store which is explained below.

3. Tools comes first and then Entertainment in most downloaded apps genres with over 8% and 6% respectively.

4. Books and refereces make the bottom of the table suggesting that android users do not prefer utitlizing ebooks over 
   android mobiles
   
5. Most of the apps downloaded are for practical purposes(e.g. Tools,Business,Education,Finance) rather than leisure(e.g. Entertainment, Social and action games)

# IOS Dataset

* As done with the android data set, we will make a frequence table for the **prime_genre** column
* We will try to observe the top 20 downloaded apps' prime genres
* And their contribution to total downloads to get a better insight on market trends

In [21]:
print('The following are valuable insights from the "Prime_Genre" column of IOS play store dataset :')
ios_primegenres_freq = freq_generator(apple_final,-5)

# Check for top 20 Categories contributions

slice_freq = ios_primegenres_freq[:20]
top_20 = 0
print('----------------------------------------------------------------')
print('Below are the top_20 Categories of most Downloaded apps with their percentages :')
for item in slice_freq:
    top_20 += float(item[0])
    print(item[1],':',item[0])

print('-----------------------------------------------------------------')
print('The top_20 apps Categories contribute to Downloads :')
print(top_20,'%')


The following are valuable insights from the "Prime_Genre" column of IOS play store dataset :
----------------------------------------------------------------
Below are the top_20 Categories of most Downloaded apps with their percentages :
Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
-----------------------------------------------------------------
The top_20 apps Categories contribute to Downloads :
99.5034140285537 %


## Takeaways from ios prime_genre frequency table

1. Game is the most common genre with over 58 % of the total downloads.
2. Entertainment comes second but there is a stark drop in percentage i.e. only 7.88% apps belong to this genre.
3. Books,Reference and Business related apps are the least downloaded.
4. Most of the apps downloaded are for leisure (e.g.Games,Entertainment,Photo and Video etc) rather than productivity.

 ## Analyzing ios Dataset wrt. user ratings
 
 * In our next step step of analysis, we are interested in seeking out the average ratings for each prime_genre in th apple data set.This will help to separate the genres which got the most user ratings.
 * That would be a good indicator to look for the market preference.
 * First of all ,we separate out all the prime_genres of ios data set and store them as  a list
 * After that we compute the average user rating for each genre, this is attained by adding up the user rating of all apps in the specific category and divided by the total number of apps in that genre.
 * We also maintain a book_keeping list to sort out the genres with descending order of average user ratings.
 * After that, we dig deep into the top two genres and seperate out the top apps with respect to user ratings
 

In [22]:
print('"Function to return the list of  genres ""')
def freq_generator(dataset,index):
    genre_freq = {}
    freq_list = []
    for row in dataset:
        genre_var = row[index]
        if genre_var in genre_freq:
            genre_freq[genre_var] += 1
        else :
            genre_freq[genre_var] = 1
    for key in genre_freq:
        freq_list.append(key)
    
    return freq_list

# calling the function to obtain list of genres

primegenre_freq = freq_generator(apple_final,-5)
print('-----------------------------------------------')
print('The total number of genres in ios Data set are :' , len(primegenre_freq) )
print('-----------------------------------------------')
print('The ios dataset has the following genres:')


print(primegenre_freq)



"Function to return the list of  genres ""
-----------------------------------------------
The total number of genres in ios Data set are : 23
-----------------------------------------------
The ios dataset has the following genres:
['Productivity', 'Weather', 'Shopping', 'Reference', 'Finance', 'Music', 'Utilities', 'Travel', 'Social Networking', 'Sports', 'Health & Fitness', 'Games', 'Food & Drink', 'News', 'Book', 'Photo & Video', 'Entertainment', 'Business', 'Lifestyle', 'Education', 'Navigation', 'Medical', 'Catalogs']


In [23]:
'''Computing average rating with respect to each genre i.e. 
   total number of user ratings for the apps that fall into the
   genre divided by the total apps in that genre'''
print('----------------------------------------------------------------------------------')
print('the average ratings wrt. to genre are')
book_keeping = [] #list to sort out prime_genres in descending order of average user ratings
len_genre = 0
total = 0
for prime in primegenre_freq:
    for row in apple_final:
        genre = row[-5]
        if genre == prime:
            total += float(row[6])
            len_genre += 1
    average = int(total/len_genre)
    book_keeping.append((average,prime))
    print(prime,':', average)
print('-----------------------------------------------------------------------------------------')
print('Genres sorted with decreasing order of  average user ratings')
book_keeping = sorted(book_keeping,reverse=True)
book_keeping


----------------------------------------------------------------------------------
the average ratings wrt. to genre are
Productivity : 21028
Weather : 31445
Shopping : 29182
Reference : 33610
Finance : 33263
Music : 38777
Utilities : 34367
Travel : 33768
Social Networking : 41544
Sports : 39354
Health & Fitness : 37746
Games : 26636
Food & Drink : 26704
News : 26614
Book : 26684
Photo & Video : 26786
Entertainment : 25713
Business : 25611
Lifestyle : 25460
Education : 24781
Navigation : 24895
Medical : 24850
Catalogs : 24824
-----------------------------------------------------------------------------------------
Genres sorted with decreasing order of  average user ratings


[(41544, 'Social Networking'),
 (39354, 'Sports'),
 (38777, 'Music'),
 (37746, 'Health & Fitness'),
 (34367, 'Utilities'),
 (33768, 'Travel'),
 (33610, 'Reference'),
 (33263, 'Finance'),
 (31445, 'Weather'),
 (29182, 'Shopping'),
 (26786, 'Photo & Video'),
 (26704, 'Food & Drink'),
 (26684, 'Book'),
 (26636, 'Games'),
 (26614, 'News'),
 (25713, 'Entertainment'),
 (25611, 'Business'),
 (25460, 'Lifestyle'),
 (24895, 'Navigation'),
 (24850, 'Medical'),
 (24824, 'Catalogs'),
 (24781, 'Education'),
 (21028, 'Productivity')]

# Takeawys from ios user ratings Analysis

1. Top two genres from the ios dataset with respect to average user ratings are 'Social Netwroking' and 'Sports'
2. The Soical networking is heavily saturated and top five apps have the monopoly over the market.So for a small company,venturing into social networking does not seem a bright idea for a short time span.

3. The sports cateogry is interesting.Top two apps make for the most of the user ratings but the saturation is not that visible unlike social networking app.There are possibilities for real time sports applications like fantasy leagues and they prove to be profitable.

4. Categories with least average user rating does not mean their demand is likey to decresase in the future.For example ,Navigation category can be exploited with better apps to create a niche for the company

## Analyzing Google Dataset wrt. user ratings

1. Now we dig into the google data set and perform the same analysis with respect to number of installs.
2. For this analysis ,we will consider the **Category** column which is the parent set of **Genre**


In [24]:
print('"Function to return the list of  Categories ""')
def freq_generator(dataset,index):
    genre_freq = {}
    freq_list = []
    for row in dataset:
        genre_var = row[index]
        if genre_var in genre_freq:
            genre_freq[genre_var] += 1
        else :
            genre_freq[genre_var] = 1
    for key in genre_freq:
        freq_list.append(key)
    
    return freq_list

# calling the function to obtain list of genres

cat_freq = freq_generator(android_final,1)
print('-----------------------------------------------')
print('The total number of Categories in google Data set are :' , len(primegenre_freq) )
print('-----------------------------------------------')
print('The google dataset has the following categories:')


print(cat_freq)


"Function to return the list of  Categories ""
-----------------------------------------------
The total number of Categories in google Data set are : 23
-----------------------------------------------
The google dataset has the following categories:
['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY', 'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION', 'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE', 'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL', 'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL', 'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER', 'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION']


In [25]:
'''Computing average number of installs with respect to each category i.e. 
   total number of installs for the apps that fall into the
   category divided by the total apps in that category'''
print('----------------------------------------------------------------------------------')
print('the average installs wrt. to category are')
book_keeping2 = [] #list to sort out categories in descending order of average installs
length = 0
total = 0
for cat in cat_freq:
    for row in android_final:
        category = row[1]
        if category == cat:
            n_rating = float(row[5].replace('+','').replace(',',''))
            total += n_rating
            length += 1
    
    average2 = float(total/length)
    book_keeping2.append((average2,cat))
    print(cat,':', average2)

print('-----------------------------------------------------------------------------------------')
print('Categories sorted with decreasing order of  average installs')
book_keeping2 = sorted(book_keeping2,reverse=True)
book_keeping2


----------------------------------------------------------------------------------
the average installs wrt. to category are
ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 1196411.2302158272
BEAUTY : 1007803.1822916666
BOOKS_AND_REFERENCE : 4867493.3795811515
BUSINESS : 3239904.3865652727
COMICS : 3082056.5296208533
COMMUNICATION : 12058498.595932802
DATING : 10632003.602623457
EDUCATION : 9984222.065046461
ENTERTAINMENT : 10079101.528975742
EVENTS : 9678965.629605688
FINANCE : 8228545.5792
FOOD_AND_DRINK : 7879225.043828716
HEALTH_AND_FITNESS : 7433042.5659876
HOUSE_AND_HOME : 7241961.636636636
LIBRARIES_AND_DEMO : 7014916.48094449
LIFESTYLE : 6315758.265942029
GAME : 8522463.352843734
FAMILY : 6995518.677614194
MEDICAL : 6612010.389948316
SOCIAL : 7283709.970925261
SHOPPING : 7275585.640919616
PHOTOGRAPHY : 7712773.045822103
SPORTS : 7527192.839437046
TRAVEL_AND_LOCAL : 7723315.387967718
TOOLS : 8028478.234368804
PERSONALIZATION : 7922722.195062985
PRODUCTIVITY : 8295502.565

[(12058498.595932802, 'COMMUNICATION'),
 (10632003.602623457, 'DATING'),
 (10079101.528975742, 'ENTERTAINMENT'),
 (9984222.065046461, 'EDUCATION'),
 (9678965.629605688, 'EVENTS'),
 (8552401.665331807, 'NEWS_AND_MAGAZINES'),
 (8523291.838789448, 'VIDEO_PLAYERS'),
 (8522463.352843734, 'GAME'),
 (8489513.914147113, 'MAPS_AND_NAVIGATION'),
 (8295502.565212091, 'PRODUCTIVITY'),
 (8241076.501452433, 'PARENTING'),
 (8228545.5792, 'FINANCE'),
 (8214096.072842914, 'WEATHER'),
 (8028478.234368804, 'TOOLS'),
 (7922722.195062985, 'PERSONALIZATION'),
 (7879225.043828716, 'FOOD_AND_DRINK'),
 (7723315.387967718, 'TRAVEL_AND_LOCAL'),
 (7712773.045822103, 'PHOTOGRAPHY'),
 (7527192.839437046, 'SPORTS'),
 (7433042.5659876, 'HEALTH_AND_FITNESS'),
 (7283709.970925261, 'SOCIAL'),
 (7275585.640919616, 'SHOPPING'),
 (7241961.636636636, 'HOUSE_AND_HOME'),
 (7014916.48094449, 'LIBRARIES_AND_DEMO'),
 (6995518.677614194, 'FAMILY'),
 (6612010.389948316, 'MEDICAL'),
 (6315758.265942029, 'LIFESTYLE'),
 (4867493.3795

In [26]:
print('Analyzing top two genres')

# function t_apps return the list of apps with most installs in descending order

def t_apps(dataset,cat):
    
    app_list = []
    for row in dataset:
            name = row[0]
            category = row[1]
            n_rating = float(row[5].replace('+','').replace(',',''))
            if category == cat:
                app_list.append((n_rating,name))
    app_list = sorted(app_list,reverse=True)
    return app_list


print('-------------------------------------------------------------')
print('The  apps in  highest averaged user rating category "Communication" is')
communication = book_keeping2[0][1]  #top category'Communication'
comm_list = t_apps(android_final,communication)
print(comm_list[5])
print('----------------------------------------------------------------')
print('The  apps in  2nd highest averaged user rating category "Dating" is')
dating = book_keeping2[1][1]
dating_list = t_apps(android_final,dating)    #top _2nd category with respect to user ratings
print(dating_list[:5])

        


Analyzing top two genres
-------------------------------------------------------------
The  apps in  highest averaged user rating category "Communication" is
(1000000000.0, 'Gmail')
----------------------------------------------------------------
The  apps in  2nd highest averaged user rating category "Dating" is
[(10000000.0, 'Zoosk Dating App: Meet Singles'), (10000000.0, 'OkCupid Dating'), (10000000.0, 'Moco - Chat, Meet People'), (10000000.0, 'Match™ Dating - Meet Singles'), (10000000.0, 'Hot or Not - Find someone right now')]


# Takeawys from Android number of installs Analysis

1. Top two categories with respect to average number of installs are communication and dating.
2. Both two categories are pretty much saturated with top 5 apps having a visible monopoly.
3. It is not advisable to venture into these categories for fast paced ad revenue.
4. Rather since from our previous analysis of total number of apps in each category and from this data,it is a nice bet
   to look into development of apps with **productivity** and **tools** category because they in the middle of most installs table and show promising future since android users prefer productivity apps more than leisure one